### ISSUE TITLE
Check the results of the energy fixer after running YOG

### ISSUE THREAD

**Issue by jatkinson1000** [m2lines/CAM-ML#34]
It would be good to see how much (if any) lifting the energy fixer is having to do after the YOG parameterisation is run.

Diagnostics should be output to the CAM log files which can be used to check this.

**Comment by jatkinson1000** [comment]
This goes to the logfiles and is output.
Judith can provide guidance but look at the `check_energy` subroutine.

**Comment by MarionBWeinzierl** [comment]
@paogorman said that this is lower priority for us, as it involves more science (might be picked up by research associate in the new year?).

**Comment by paogorman** [comment]
Yes, this could be a lower priority than other issues. One caveat: checking the energy fixer for YOG could be a way to find a bug in the code if it flagged that energy was clearly not conserved.

**Comment by gretamiller** [comment]
Yes, Jack pointed out the line I think Judith was referring to: https://github.com/m2lines/CAM-ML/blob/4a4a0a6d8982c2f3d32bbda421e585fc1daa0c5e/src/physics/cam/physpkg.F90#L1975 ...but I'm not sure about log files for this

**Comment by MarionBWeinzierl** [comment]
Information about the dycore energy fixer can be found here: https://www.cesm.ucar.edu/models/simple/change-dycore . @judithberner mentioned, though, that we are only interested in the physics energy fixer.

**Comment by MarionBWeinzierl** [comment]
Doing a `grep` of `EFIX`, as was mentioned at the above link, shows that this field is output in the `logs/atm.log*` files.

**Comment by MarionBWeinzierl** [comment]
I output the EFIX field for the YOG simulation:
![Image](https://github.com/user-attachments/assets/8aa208e3-0935-411b-89d9-e64c8acd299c)

Is this the kind of output we would expect?

**Comment by paogorman** [comment]
Thanks Marion.

That's a larger energy correction than I would expect for a GCM, but it
would be a good idea to try it with the ZM configuration to see what you
get for comparison.

ps: Just check: Is this the output of the energy fixer for all processes?
(not just YOG)

**Comment by MarionBWeinzierl** [comment]
This is without the bugfix that Jack has done, btw. I believe it is for all processes -- they seem to all write to the same EFIX field.

This is the file with axis etc:
![Image](https://github.com/user-attachments/assets/b73744ed-dab6-4c81-97e9-faa507594994)

**Comment by paogorman** [comment]
Hi Marion,

I'm not completely sure, but I think the EFIX you are outputting is related
to the finite volume dynamical core rather than the parameterizations.
Apparently the FV core has about a 2 W/m^2 energy loss due to
discretization of the momentum equations.

To get the code to write out information about the energy conservation of
YOG, I think we need to set the parameter "print_energy_errors" in the
check_energy_nl namelist to .true.
You can see how this information is written out by looking at the
check_energy_chng subroutine in check_energy.f90.

Also, in the code below in physpkg.f90, I would change "chkengyfix" to
"yog_nn" so that what is written out to the logfiles has a name to identify
yog.

if (yog_scheme=='on') then
        call t_startf('yog_nn')
        call yog_tend(ztodt, state, ptend, pbuf)
        call physics_update(state, ptend, ztodt, tend)
        call t_stopf('yog_nn')

        flx_cnd(:ncol) = prec_dp(:ncol)
        call check_energy_chng(state, tend, "chkengyfix", nstep, ztodt,
zero, flx_cnd, zero, flx_heat)
    end if

**Comment by jatkinson1000** [comment]
From a look into the source of `check_energy_chng` I believe one needs to set a namelist variable `print_energy_errors = .true.` to enable printing of any columns violating the energy fixer tolerance to the logfile (`iulog`). This will include the name of the check where it occurs which is patched in #51

**Comment by MarionBWeinzierl** [comment]
This is the type of output I get with `print_energy_errors=true`:

![image](https://github.com/user-attachments/assets/e9b6f4b6-f806-4d2e-81d0-9f849c52e4c3)

**Comment by jatkinson1000** [comment]
Interesting, seems it's a few places that raise issues, so we may want to try and filter by occurrences of `yog_nn`.
Might be worth speaking to @tommelt as I recall he had some useful macros for reorganising MPI output text files.

**Comment by MarionBWeinzierl** [comment]
I reran the above after #51 was merged in, an all the occurrences are `yog_nn`.

**Comment by MarionBWeinzierl** [comment]
I get this for the ZM run, but there are much less of those warnings:
![image](https://github.com/user-attachments/assets/b7262589-3382-43ab-b111-2acd90770ba3)

**Comment by MarionBWeinzierl** [comment]
Looking at the output code here https://github.com/m2lines/CAM-ML/blob/625f3f33f78d9426146f9ec5a0fe57997aa81498/src/physics/cam/check_energy.F90#L476C1-L482C19 we are looking at the bottom right value for `te_rer` . The YOG version shows a significantly higher value there.

**Comment by paogorman** [comment]
Thanks Marion. Could you please attach a file showing the output when
yog_nn is used as the ‘name’? (I may have missed it)

**Comment by MarionBWeinzierl** [comment]
Oh, sorry, it's the same as above with `chkengyfix`, I just didn't repost it after the rename. Here is an extract: 
![image](https://github.com/user-attachments/assets/d5cf68ae-fbe8-4d6d-8b0f-5a5ef3c7fb7f)

The same type of messages goes over thousands of lines.

**Comment by paogorman** [comment]
Great - thanks.

It seems the water and energy conservation is not working as well in yog as
in zm. These may be fairly big errors because it is a percent of the total
water or energy (rather than a percent of the change).

It might make sense to focus on the water conservation problem first
because it is simpler. Also, if we don't conserve water we won't conserve
energy because of the particular energy that CAM uses which is
frozen_moist_enthalpy + latent_fusion*q_total - I've never seen that used
before but it is valid as long as total water is conserved.

For energy, I think we could improve a bit by predicting the snow. That
would involve a few lines to add to nn_convection_flux which I could work
out next week. But I'm guessing there is a bigger problem given that water
is not conserved.

**Comment by jatkinson1000** [comment]
Check integral of:
- vap, ice, cld in from CAM
- q in SAM
- vap, ice, liq, precip going out to CAM

Beware need to multiply by density to get appropriate mass - moist vs dry
Beware we need to integrate up to the appropriate height that matches the CAM cell in 
Also check that there is no change in water variables above the parameterisation where no tendencies are predicted.

**Comment by MarionBWeinzierl** [comment]
Following @jatkinson1000 's suggestion I am looking at Toga II as an easier example. 

YOG ( end of that file, more warnings further up):
![image](https://github.com/user-attachments/assets/6af2b05e-cd8b-44fc-97d9-b67ef972dc04)

No warnings for ZM.

Note that for the aquaplanet, there are `mixing ratio violated` warnings for both ZM and YOG.
